### PR TITLE
[reason/en] Add Reason language

### PR DESCRIPTION
**Pull request by Darth-Knoppix**
- [x] I solemnly swear that this is all original content of which I am the original author
- [x] Pull request title is prepended with `[language/lang-code]`
- [x] Pull request touches only one file (or a set of logically related files with similar changes made)
- [x] Content changes are aimed at *intermediate to experienced programmers* (this is a poor format for explaining fundamental programming concepts)
- [x] If you've changed any part of the YAML Frontmatter, make sure it is formatted according to [CONTRIBUTING.md](https://github.com/adambard/learnxinyminutes-docs/blob/master/CONTRIBUTING.markdown)
  - [x] Yes, I have double-checked quotes and field names!

### CODE REVIEW

**Comment by Darth-Knoppix** [comment]
Hi @divayprakash, I created a new language tutorial for the Reason language.
Is there anything I can do to get this reviewed or changes required?

File: reason.html.markdown [hunk 1]
```diff
@@ -0,0 +1,545 @@
+---
+name: Reason
+category: language
```
Comment (divayprakash) [inline reason.html.markdown]: Please remove these 2 lines -
```
name: Reason
category: language
```

**Comment by divayprakash** [comment]
Closes #3108

**Comment by divayprakash** [comment]
@Darth-Knoppix Thank you for the contribution and welcome to the LXIYM community! Sorry about the delay in the review, I was on an extended break.

File: reason.html.markdown [hunk 2]
```diff
@@ -0,0 +1,545 @@
...
+
+/* Alternatively, a try block can be used */
+let messageToEvan =
+  try (driveToTown(evan)) {
+  | Under_Age => evan.firstName ++ " is too young to drive!"
+  };
+/* - With Reason exceptions can be avoided with optionals and are seldom used */
+
+/*----------------------------------------------
+ * Object
```
Comment (amartincolby) [inline reason.html.markdown]: Objects in Reason are such strange SOBs that I think greater elucidation is needed. You should also add the elision operator of double dots.

File: reason.html.markdown [hunk 3]
```diff
@@ -0,0 +1,545 @@
...
+};
+
+house#temperature; /* - : float = 18. */
+house#arriveHome();
+house#temperature; /* - : float = 22. */
+
+/*----------------------------------------------
+ * Module
+ *----------------------------------------------
+ * Modules are used to organize your code and provide namespacing,
```
Comment (amartincolby) [inline reason.html.markdown]: Period at the end.

File: reason.html.markdown [hunk 4]
```diff
@@ -0,0 +1,545 @@
...
+let messageToEvan =
+  try (driveToTown(evan)) {
+  | Under_Age => evan.firstName ++ " is too young to drive!"
+  };
+/* - With Reason exceptions can be avoided with optionals and are seldom used */
+
+/*----------------------------------------------
+ * Object
+ *----------------------------------------------
+ * Objects are similar to Record types but aren't as rigid
```
Comment (amartincolby) [inline reason.html.markdown]: Period at the end.

File: reason.html.markdown [hunk 5]
```diff
@@ -0,0 +1,545 @@
...
+
+  let getMeetingTime = staff =>
+    switch (staff) {
+    | Other => 11_15 /* - : int = 1115; Underscores are for formatting only  */
+    | _ => 9_30
+    };
+};
+```
+
+## Further Reading
```
Comment (amartincolby) [inline reason.html.markdown]: Thank you for writing all of this. I'm a big fan of Reason.

**Comment by divayprakash** [comment]
@amartincolby Please feel free to make the changes you suggested in a separate PR.

File: reason.html.markdown [hunk 6]
```diff
@@ -0,0 +1,545 @@
...
+language: reason
+filename: reason.re
+contributors:
+    - ["Seth Corker", "https://sethcorker.com"]
+---
+
+Reason is a syntax over OCaml that is easier to get started for programmers who are familiar with C-style syntax like JavaScript. BuckleScript is part of the toolchain which compiles Reason to JavaScript so you can write statically typed code for anywhere that JavaScript runs.
+
+```javascript
+/* Comments start with slash-star, and end with slash-star */
```
[the remark this answers is not in the document]
Comment (Darth-Knoppix) [reply reason.html.markdown]: Thanks for identifying these changes, I've created a [PR here](https://github.com/adambard/learnxinyminutes-docs/pull/3605#issue-307096122). Feel free to make some changes or add more suggestions.

File: reason.html.markdown (hunk 5 above)
Comment (Darth-Knoppix) [reply reason.html.markdown]: No problem, I had fun doing it.

File: reason.html.markdown [hunk 7]
```diff
@@ -0,0 +1,545 @@
...
+
+/*----------------------------------------------
+ * Variable and function declaration
+ *----------------------------------------------
+ * Variables and functions use the let keyword and end with a semi-colon
+ * `let` bindings are immutable
+ */
+
+let x = 5;
+/* - Notice we didn't add a type, Reason will infer it's an int */
```
[the remark this answers is not in the document]
Comment (Darth-Knoppix) [reply reason.html.markdown]: Good suggestion, [here it is](https://github.com/adambard/learnxinyminutes-docs/pull/3605/commits/91f1f4f936f13e7363d832771acb6f5b2cd14879#diff-4011e3b30e363503b76fcbd47506bafaR21)

**Pull request merged**